esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -0,0 +1,202 @@
+/*##############################################################################
+
+HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+############################################################################## */
+
+#include "ldapenvironmentService.hpp"
+#include "exception_util.hpp"
+#include "ldapsecurity.ipp"
+
+#define MSG_SEC_MANAGER_IS_NULL "LDAP Security Manager is not found. Please check if the system authentication is set up correctly"
+#define MSG_SEC_MANAGER_ISNT_LDAP "LDAP Security Manager is required for this feature. Please enable LDAP in the system configuration"
+
+void CldapenvironmentEx::init(IPropertyTree *_cfg, const char *_process, const char *_service)
+{
+    cfg = _cfg;
+    try
+    {
+#ifdef _DEBUG
+        StringBuffer sb;
+        toXML(cfg, sb);
+        if (!sb.isEmpty())
+            DBGLOG("\n------ldapenvironment config\n%s\n------\n", sb.str());
+#endif
+        VStringBuffer prefix("Software/EspProcess[@name='%s']/@", _process);
+        StringBuffer xpath;
+
+        xpath.set(prefix).append("hpccRootName");
+        cfg->getProp(xpath.str(), ldapRootOU);
+        if (ldapRootOU.isEmpty())
+            throw MakeStringException(-1, "hpccRootName must be specified in configuration (ex. 'hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%sadminGroupName", prefix.str());
```

Comment:
Sorry, I meant you could use the xpath.set(prefix).append("hpccRootName"); form for all of these cases.

Reply:
Done